src/blockchain-utilities/derivation.ts

```diff
@@ -0,0 +1,109 @@
+import { ethers } from 'ethers';
+
+import { addressSchema } from './schema';
+
+export const PROTOCOL_IDS = {
+  RRP: '1',
+  PSP: '2',
+  RELAYED_RRP: '3',
+  RELAYED_PSP: '4',
+  AIRSEEKER: '5',
+};
+
+/**
+ * An interface that reflects the structure of a Template
+ */
+export interface Template {
+  airnode: string;
+  encodedParameters: string;
+  endpointId: string;
+}
+
+/**
+ * Derives a template ID from the input parameters
+ *
+ * @param airnode an Airnode address
+ * @param encodedParameters encoded parameters, see the airnode/abi package's encode function
+ * @param endpointId An endpointID (see deriveEndpointId)
+ */
+export const deriveTemplateId = ({ airnode, encodedParameters, endpointId }: Template) =>
+  ethers.utils.solidityKeccak256(['address', 'bytes32', 'bytes'], [airnode, endpointId, encodedParameters]);
+
+/**
+ * Derives an endpoint ID
+ *
+ * @param oisTitle the OIS title
+ * @param endpointName the endpoint name
+ */
+export const deriveEndpointId = (oisTitle: string, endpointName: string) =>
+  ethers.utils.keccak256(ethers.utils.defaultAbiCoder.encode(['string', 'string'], [oisTitle, endpointName]));
+
+/**
+ * Derives an airnode address's xpub, required for allowing signed data consumers to verify authenticity
+ *
+ * @param airnodeMnemonic the airnode's mnemonic
+ */
+export const deriveAirnodeXpub = (airnodeMnemonic: string) =>
+  ethers.utils.HDNode.fromMnemonic(airnodeMnemonic).derivePath("m/44'/60'/0'").neuter().extendedKey;
+
+/**
+ * Derives a wallet path from a sponsor address, used for calculating a sponsor wallet.
+ *
+ * @param sponsorAddress an EVM-compatible address
+ * @param protocolId an API application protocol ID
+ */
+export function deriveWalletPathFromSponsorAddress(sponsorAddress: string, protocolId: string) {
+  addressSchema.parse(sponsorAddress);
+
+  const sponsorAddressBN = ethers.BigNumber.from(sponsorAddress);
+  const paths = [];
+  for (let i = 0; i < 6; i++) {
+    const shiftedSponsorAddressBN = sponsorAddressBN.shr(31 * i);
+    paths.push(shiftedSponsorAddressBN.mask(31).toString());
+  }
+  return `${protocolId}/${paths.join('/')}`;
+}
+
+/**
+ * Encodes/formats a string as a hex-encoded bytes32 string.
+ *
+ * @param input the input string
+ */
+export const formatBytes32String = (input: string) => ethers.utils.formatBytes32String(input);
```

Comment:
Can we call it `toBytes32String` and also have the counterpart `fromBytes32String` ?

Reply:
Done 💪